### PR TITLE
Sorts the insides of the Par and ESet/EMap of the rholangADT

### PR DESCRIPTION
**Pull request by KentShikama**
A score tree is recursively built for each term and is used to sort the insides of Par/ESet/EMap. For most terms, the current term type's absolute value based on the Score object is added as a Leaf to the left most branch and the score tree built for the inside terms are added to the right. The Score object is a container of constants that arbitrarily assigns absolute values to term types. The sort order is total as every term type is assigned an unique value in the Score object. For ground types, the appropriate integer representation is used as the base score tree. For var types, the Debruijn level from the normalization is used.

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+  // Shortcut to be able to write Leaves(1,2,3) instead of Node(Seq(Leaf(1),Leaf(2),Leaf(3)))
+  def apply[T](children: T*) : Node[T] = Node(children.map(a => Leaf(a)))
+}
+
+// Effectively a tuple that groups the term to its score tree.
+case class ScoredTerm[T](term: T, score: Tree[Int]) extends Ordered[ScoredTerm[T]] {
+  def compare(that: ScoredTerm[T]) : Int = {
+    def compareScore(s1: Tree[Int], s2: Tree[Int]) : Int = {
+      (s1, s2) match {
+        case (Leaf(a), Leaf(b)) => a.compare(b)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Make the tree a tree of Either[Int, String]. Make sure there's a sort order for Either. If not, create your own case class and add a sort order.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+  def sortBind(bind: (List[Channel], Channel)): ScoredTerm[Tuple2[List[Channel], Channel]] = {
+    val (patterns, channel) = bind
+    val sortedPatterns = patterns.map(channel => ChannelSortMatcher.sortMatch(channel))
+    val sortedChannel = ChannelSortMatcher.sortMatch(channel)
+    ScoredTerm((sortedPatterns.map(_.term), sortedChannel.term),
+                Node(Seq(sortedChannel.score) ++ sortedPatterns.map(_.score)))
+  }
+
+  def sortMatch(r: Receive) : ScoredTerm[Receive] = {
+    val sortedBinds = r.binds.map(bind => sortBind(bind))
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Sorting the binds should be as simple as adding .sorted to this line.

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Per our conversation, this function should not sort the binds.

Instead, write a helper method that can sort a list of (Channel, List[Channel], DebruijnLevelMap) triples

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala [hunk 3]
```diff
@@ -270,6 +269,20 @@ object ReceiveSortMatcher {
                 Node(Seq(sortedChannel.score) ++ sortedPatterns.map(_.score)))
   }
 
+  // Used during normalize to presort the binds.
+  def preSortBinds[T](binds: List[Tuple3[List[Channel], Channel, DebruijnLevelMap[T]]]) :
+    List[Tuple3[List[Channel], Channel, DebruijnLevelMap[T]]] = {
+    val sortedBind = binds.map{
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: space after map

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala (hunk 1 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    }
+  }
+}
+
+object GroundSortMatcher {
+  def sortMatch(g: Ground) : ScoredTerm[Ground] = {
+    g match {
+      case gb: GBool => ScoredTerm(g, BoolSortMatcher.sortMatch(gb).score)
+      case gi: GInt => ScoredTerm(g, Leaves(Score.INT, gi.i))
+      // TODO: Think about hashCode collisions
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala (hunk 3 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala]: Done

**Comment by KentShikama** [comment]
@guardbotmk3 Ready for another round of reviews.

**Pull request merged**